transit_odp/timetables/transformers.py

```diff
@@ -181,3 +283,81 @@ def sync_stop_points(self, stop_points, provisional_stops):
         # Return the subselection of stop points seen in the doc (useful when
         # processing large zip files)
         return stop_point_cache.reindex(sorted(stop_point_refs))
+
+    def sync_flexible_zone(self, flexible_stop_points):
+        filtered_stop_points = flexible_stop_points[
+            (flexible_stop_points["bus_stop_type"] == "flexible")
+            & (~flexible_stop_points["naptan_id"].isna())
+        ]
+        filtered_stop_points_naptan_id = filtered_stop_points["naptan_id"].tolist()
+        flexiblezone_qs = FlexibleZone.objects.filter(
+            naptan_stoppoint_id__in=filtered_stop_points_naptan_id
+        )
+        flexible_zone_df = create_naptan_flexible_zone_df_from_queryset(flexiblezone_qs)
+        if not flexible_zone_df.empty:
+            filtered_stop_points_naptan_id = filtered_stop_points.reset_index().merge(
+                flexible_zone_df.reset_index(),
+                how="left",
+                left_on="naptan_id",
+                right_on="naptan_id",
+            )
+            filtered_stop_points_naptan_id = filtered_stop_points_naptan_id[
+                ["atco_code", "flexible_location"]
+            ]
+            filtered_stop_points_naptan_id.set_index("atco_code", inplace=True)
+            return filtered_stop_points_naptan_id
+        return pd.DataFrame()
+
+    def create_flexible_timing_link(self, flexible_journey_details):
+        def get_stop_sequence(group, journey_pattern_id):
+            group["order"] = range(len(group))
+            group["journey_pattern_id"] = journey_pattern_id
+            group["from_stop_atco"] = group["atco_code"].shift(0)
+            group["to_stop_atco"] = group["atco_code"].shift(-1)
+            return group.dropna(subset=["to_stop_atco"])
+
+        if not flexible_journey_details.empty:
+            grouped = flexible_journey_details.reset_index().groupby(
```

Comment:
We could pass the function get_stop_sequence as an apply function after groupby to return expected data to avoid for iteration below. transform_service_pattern_stop can be used to for reference

Reply:
agree